themes/classic/_dev/js/components/product-miniature.js

```diff
@@ -28,24 +28,34 @@ export default class ProductMinitature {
   init() {
     $('.js-product-miniature').each((index, element) => {
       const FLAG_MARGIN = 10;
-      const discountElems = $(element).find('.discount-product');
+      //Top on sale banner
       const onSaleElems =  $(element).find('.on-sale');
-      const newElems = $(element).find('.new');
-
-      if (discountElems.length) {
-        newElems.css('top', discountElems.height() * 3 + FLAG_MARGIN);
-        discountElems.css('top', -$(element).find('.thumbnail-container').height() + $(element).find('.product-description').height() + FLAG_MARGIN);
+      //Discount flag
+      const discountElems = $(element).find('.discount-product');
+      //Flags other than on-sale, discount and online-only (which have all their way to display)
+      const flagElems = $(element).find('.product-flag:not(.on-sale):not(.discount):not(.online-only)');
 
-        if ($(element).find('.pack').length) {
-          $(element).find('.pack').css('top', discountElems.height() * 3 + FLAG_MARGIN);
-        }
+      var flagsTop = FLAG_MARGIN;
+      var discountTop = FLAG_MARGIN;
+      if (onSaleElems.length) {
+        discountTop = onSaleElems.outerHeight() + FLAG_MARGIN;
       }
 
-      if (onSaleElems.length) {
-        discountElems.css('top', parseFloat(discountElems.css('top')) + onSaleElems.height() + FLAG_MARGIN);
-        newElems.css('top', (discountElems.height() * 3 + onSaleElems.height()) + FLAG_MARGIN * 3);
+      if (discountElems.length) {
+        flagsTop = discountTop + discountElems.outerHeight() + FLAG_MARGIN;
+        //Discount flag is actually in product-description div so it needs a negative top value
+        discountElems.css('top', discountTop + -$(element).find('.thumbnail-container').outerHeight() + $(element).find('.product-description').outerHeight());
       }
 
+      //Now display flags one above the other
+      flagElems.each((index, flag) => {
+        $(flag).css('top', flagsTop);
+        $(flag).css('margin-top', 0);
```

Comment:
Should we define this rule in a CSS file instead?

Reply:
Actually there is already a non 0 rule in the css, but I didn't want to modify it This way if javascript has problems or in pages where this code doesn't run the css rule is still there as a fallback
Maybe I should add a comment however here?

Reply:
Well, if it is already in the css file, I see no reason to redefine it here again.

This could have unexpected consequences for child themes, who may prefer to set their label on another place of the website. If your changes work even without this line, I would rather remove it.

Reply:
That's the thing it doesn't work correctly without this line, because the margin from the css offsets the elements. Since the `top` value is computed by the js the `margin-top` is not really useful any more.
I can remove the line from the css but it could also have consequences on a child theme no? Honestly I don't even understand why this is performed through javascript, a full css method would probably be better but I only fixed the existing one. Maybe dropping this js computing with a css layout would be better.

Reply:
It's best just to add a class and avoid setting styles as much as possible

Reply:
I agree, I'll rework this feature to be full css then and drop the javascript as much as possible.
But as special product flags may have very different looks some js might still be necessary

Reply:
Alright! no more js positioning And I also cleaned CSS a little as the product-flag were defined in two places without valid reason (for product page and product list page)